File: visolingua-rust/src-tauri/src/translator.rs
use crate::config::ConfigData;
use anyhow::{Context, Result};
use reqwest::Client;
use serde_json::{json, Value};

/// Translate an image using the configured LLM
pub async fn translate_image_with_data(config: &ConfigData, image_base64: &str) -> Result<String> {
    let llm_provider = &config.llm_provider;
    let target_language = &config.target_language;

    match llm_provider.as_str() {
        "gemini" => translate_with_gemini(config, image_base64, target_language).await,
        "openai" => translate_with_openai(config, image_base64, target_language).await,
        "ollama" => translate_with_ollama(config, image_base64, target_language).await,
        _ => anyhow::bail!("Unknown LLM provider: {}", llm_provider),
    }
}

/// Ask AI a question about the translation
pub async fn ask_ai_with_data(config: &ConfigData, question: &str, context: &str) -> Result<String> {
    let llm_provider = &config.llm_provider;

    let prompt = format!(
        "Based on this translation:\n\n{}\n\nUser question: {}\n\nProvide a helpful answer.",
        context, question
    );

    match llm_provider.as_str() {
        "gemini" => ask_gemini(config, &prompt).await,
        "openai" => ask_openai(config, &prompt).await,
        "ollama" => ask_ollama(config, &prompt).await,
        _ => anyhow::bail!("Unknown LLM provider: {}", llm_provider),
    }
}

async fn translate_with_gemini(config: &ConfigData, image_base64: &str, target_lang: &str) -> Result<String> {
    let api_key = &config.gemini_api_key;
    if api_key.is_empty() {
        anyhow::bail!("Gemini API key not configured");
    }
    let model = &config.gemini_model;

    let client = Client::new();
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{}:generateContent?key={}",
        model, api_key
    );

    let body = json!({
        "contents": [{
            "parts": [
                {
                    "text": format!(
                        "Translate the text in this image to {}. Only provide the translation, no explanations.",
                        target_lang
                    )
                },
                {
                    "inline_data": {
                        "mime_type": "image/png",
                        "data": image_base64
                    }
                }
            ]
        }]
    });

    let response = client
        .post(&url)
        .json(&body)
        .send()
        .await
        .context("Failed to send request to Gemini")?;

    let json: Value = response
        .json()
        .await
        .context("Failed to parse Gemini response")?;

    extract_gemini_text(&json)
}

async fn translate_with_openai(config: &ConfigData, image_base64: &str, target_lang: &str) -> Result<String> {
    let api_key = &config.openai_api_key;
    if api_key.is_empty() {
        anyhow::bail!("OpenAI API key not configured");
    }
    let model = &config.openai_model;

    let client = Client::new();
    let url = "https://api.openai.com/v1/chat/completions";

    let body = json!({
        "model": model,
        "messages": [{
            "role": "user",
            "content": [
                {
                    "type": "text",
                    "text": format!(
                        "Translate the text in this image to {}. Only provide the translation, no explanations.",
                        target_lang
                    )
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": format!("data:image/png;base64,{}", image_base64)
                    }
                }
            ]
        }],
        "max_tokens": 1000
    });

    let response = client
        .post(url)
        .header("Authorization", format!("Bearer {}", api_key))
        .json(&body)
        .send()
        .await
        .context("Failed to send request to OpenAI")?;

    let json: Value = response
        .json()
        .await
        .context("Failed to parse OpenAI response")?;

    extract_openai_text(&json)
}

async fn translate_with_ollama(config: &ConfigData, image_base64: &str, target_lang: &str) -> Result<String> {
    let base_url = &config.ollama_url;
    let model = &config.ollama_model;

    let client = Client::new();
    let url = format!("{}/api/generate", base_url);

    let body = json!({
        "model": model,
        "prompt": format!(
            "Translate the text in this image to {}. Only provide the translation, no explanations.",
            target_lang
        ),
        "images": [image_base64],
        "stream": false
    });

    let response = client
        .post(&url)
        .json(&body)
        .send()
        .await
        .context("Failed to send request to Ollama")?;

    let json: Value = response
        .json()
        .await
        .context("Failed to parse Ollama response")?;

    json["response"]
        .as_str()
        .map(|s| s.to_string())
        .context("Invalid Ollama response")
}

async fn ask_gemini(config: &ConfigData, prompt: &str) -> Result<String> {
    let api_key = &config.gemini_api_key;
    if api_key.is_empty() {
        anyhow::bail!("Gemini API key not configured");
    }
    let model = &config.gemini_model;

    let client = Client::new();
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{}:generateContent?key={}",
        model, api_key
    );

    let body = json!({
        "contents": [{
            "parts": [{"text": prompt}]
        }]
    });

    let response = client
        .post(&url)
        .json(&body)
        .send()
        .await
        .context("Failed to send request to Gemini")?;

    let json: Value = response
        .json()
        .await
        .context("Failed to parse Gemini response")?;

    extract_gemini_text(&json)
}

async fn ask_openai(config: &ConfigData, prompt: &str) -> Result<String> {
    let api_key = &config.openai_api_key;
    if api_key.is_empty() {
        anyhow::bail!("OpenAI API key not configured");
    }
    let model = &config.openai_model;

    let client = Client::new();
    let url = "https://api.openai.com/v1/chat/completions";

    let body = json!({
        "model": model,
        "messages": [{
            "role": "user",
            "content": prompt
        }],
        "max_tokens": 1000
    });

    let response = client
        .post(url)
        .header("Authorization", format!("Bearer {}", api_key))
        .json(&body)
        .send()
        .await
        .context("Failed to send request to OpenAI")?;

    let json: Value = response
        .json()
        .await
        .context("Failed to parse OpenAI response")?;

    extract_openai_text(&json)
}

async fn ask_ollama(config: &ConfigData, prompt: &str) -> Result<String> {
    let base_url = &config.ollama_url;
    let model = &config.ollama_model;

    let client = Client::new();
    let url = format!("{}/api/generate", base_url);

    let body = json!({
        "model": model,
        "prompt": prompt,
        "stream": false
    });

    let response = client
        .post(&url)
        .json(&body)
        .send()
        .await
        .context("Failed to send request to Ollama")?;

    let json: Value = response
        .json()
        .await
        .context("Failed to parse Ollama response")?;

    json["response"]
        .as_str()
        .map(|s| s.to_string())
        .context("Invalid Ollama response")
}

fn extract_gemini_text(json: &Value) -> Result<String> {
    json["candidates"][0]["content"]["parts"][0]["text"]
        .as_str()
        .map(|s| s.to_string())
        .context("Invalid Gemini response format")
}

fn extract_openai_text(json: &Value) -> Result<String> {
    json["choices"][0]["message"]["content"]
        .as_str()
        .map(|s| s.to_string())
        .context("Invalid OpenAI response format")
}

File: visolingua-go/translator.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

type Translator struct {
	config *Config
	client *http.Client
}

func NewTranslator(config *Config) *Translator {
	return &Translator{
		config: config,
		client: &http.Client{},
	}
}

// TranslateImage translates text in an image
func (t *Translator) TranslateImage(imageBase64 string) (string, error) {
	switch t.config.LLM.Provider {
	case "gemini":
		return t.translateWithGemini(imageBase64)
	case "openai":
		return t.translateWithOpenAI(imageBase64)
	case "ollama":
		return t.translateWithOllama(imageBase64)
	default:
		return "", fmt.Errorf("unknown LLM provider: %s", t.config.LLM.Provider)
	}
}

// AskAI asks a question about the translation
func (t *Translator) AskAI(question, context string) (string, error) {
	prompt := fmt.Sprintf("Based on this translation:\n\n%s\n\nUser question: %s\n\nProvide a helpful answer.", context, question)

	switch t.config.LLM.Provider {
	case "gemini":
		return t.askGemini(prompt)
	case "openai":
		return t.askOpenAI(prompt)
	case "ollama":
		return t.askOllama(prompt)
	default:
		return "", fmt.Errorf("unknown LLM provider: %s", t.config.LLM.Provider)
	}
}

func (t *Translator) translateWithGemini(imageBase64 string) (string, error) {
	if t.config.APIKeys.Gemini == "" {
		return "", fmt.Errorf("Gemini API key not configured")
	}

	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/%s:generateContent?key=%s",
		t.config.LLM.GeminiModel, t.config.APIKeys.Gemini)

	payload := map[string]interface{}{
		"contents": []map[string]interface{}{
			{
				"parts": []map[string]interface{}{
					{
						"text": fmt.Sprintf("Translate the text in this image to %s. Only provide the translation, no explanations.", t.config.Translation.TargetLanguage),
					},
					{
						"inline_data": map[string]string{
							"mime_type": "image/png",
							"data":      imageBase64,
						},
					},
				},
			},
		},
	}

	body, _ := json.Marshal(payload)
	fmt.Printf("Gemini request URL: %s\n", url)

	resp, err := t.client.Post(url, "application/json", bytes.NewBuffer(body))
	if err != nil {
		fmt.Printf("Gemini request error: %v\n", err)
		return "", fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	// Read response body
	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Printf("Failed to read response: %v\n", err)
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	fmt.Printf("Gemini response status: %d\n", resp.StatusCode)
	fmt.Printf("Gemini response: %s\n", string(responseBody))

	if resp.StatusCode != 200 {
		return "", fmt.Errorf("API error (%d): %s", resp.StatusCode, string(responseBody))
	}

	var result map[string]interface{}
	if err := json.Unmarshal(responseBody, &result); err != nil {
		fmt.Printf("JSON parse error: %v\n", err)
		return "", fmt.Errorf("failed to parse JSON: %w", err)
	}

	// Extract text from response with error checking
	candidates, ok := result["candidates"].([]interface{})
	if !ok || len(candidates) == 0 {
		return "", fmt.Errorf("no candidates in response")
	}

	candidate0, ok := candidates[0].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid candidate format")
	}

	content, ok := candidate0["content"].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid content format")
	}

	parts, ok := content["parts"].([]interface{})
	if !ok || len(parts) == 0 {
		return "", fmt.Errorf("no parts in content")
	}

	part0, ok := parts[0].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid part format")
	}

	text, ok := part0["text"].(string)
	if !ok {
		return "", fmt.Errorf("no text in part")
	}

	return text, nil
}

func (t *Translator) translateWithOpenAI(imageBase64 string) (string, error) {
	if t.config.APIKeys.OpenAI == "" {
		return "", fmt.Errorf("OpenAI API key not configured")
	}

	url := "https://api.openai.com/v1/chat/completions"

	payload := map[string]interface{}{
		"model": t.config.LLM.OpenAIModel,
		"messages": []map[string]interface{}{
			{
				"role": "user",
				"content": []map[string]interface{}{
					{
						"type": "text",
						"text": fmt.Sprintf("Translate the text in this image to %s. Only provide the translation, no explanations.", t.config.Translation.TargetLanguage),
					},
					{
						"type": "image_url",
						"image_url": map[string]string{
							"url": fmt.Sprintf("data:image/png;base64,%s", imageBase64),
						},
					},
				},
			},
		},
		"max_tokens": 1000,
	}

	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", t.config.APIKeys.OpenAI))

	resp, err := t.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	choices := result["choices"].([]interface{})
	message := choices[0].(map[string]interface{})["message"].(map[string]interface{})
	return message["content"].(string), nil
}

func (t *Translator) translateWithOllama(imageBase64 string) (string, error) {
	url := fmt.Sprintf("%s/api/generate", t.config.LLM.OllamaURL)

	payload := map[string]interface{}{
		"model":  t.config.LLM.OllamaModel,
		"prompt": fmt.Sprintf("Translate the text in this image to %s. Only provide the translation, no explanations.", t.config.Translation.TargetLanguage),
		"images": []string{imageBase64},
		"stream": false,
	}

	body, _ := json.Marshal(payload)
	resp, err := t.client.Post(url, "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	return result["response"].(string), nil
}

func (t *Translator) askGemini(prompt string) (string, error) {
	if t.config.APIKeys.Gemini == "" {
		return "", fmt.Errorf("Gemini API key not configured")
	}

	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/%s:generateContent?key=%s",
		t.config.LLM.GeminiModel, t.config.APIKeys.Gemini)

	payload := map[string]interface{}{
		"contents": []map[string]interface{}{
			{
				"parts": []map[string]interface{}{
					{"text": prompt},
				},
			},
		},
	}

	body, _ := json.Marshal(payload)
	resp, err := t.client.Post(url, "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	var result map[string]interface{}
	json.Unmarshal(respBody, &result)

	candidates := result["candidates"].([]interface{})
	content := candidates[0].(map[string]interface{})["content"].(map[string]interface{})
	parts := content["parts"].([]interface{})
	return parts[0].(map[string]interface{})["text"].(string), nil
}

func (t *Translator) askOpenAI(prompt string) (string, error) {
	if t.config.APIKeys.OpenAI == "" {
		return "", fmt.Errorf("OpenAI API key not configured")
	}

	url := "https://api.openai.com/v1/chat/completions"

	payload := map[string]interface{}{
		"model": t.config.LLM.OpenAIModel,
		"messages": []map[string]interface{}{
			{
				"role":    "user",
				"content": prompt,
			},
		},
		"max_tokens": 1000,
	}

	body, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", t.config.APIKeys.OpenAI))

	resp, err := t.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	choices := result["choices"].([]interface{})
	message := choices[0].(map[string]interface{})["message"].(map[string]interface{})
	return message["content"].(string), nil
}

func (t *Translator) askOllama(prompt string) (string, error) {
	url := fmt.Sprintf("%s/api/generate", t.config.LLM.OllamaURL)

	payload := map[string]interface{}{
		"model":  t.config.LLM.OllamaModel,
		"prompt": prompt,
		"stream": false,
	}

	body, _ := json.Marshal(payload)
	resp, err := t.client.Post(url, "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)

	return result["response"].(string), nil
}